### PR TITLE
Fix Sega CD/Saturn disc hashing

### PR DESCRIPTION
**Pull request by kalimag**
Fix `DiscHasher.OldHash()` method used to hash non-PSX discs
- Don't hash generated (empty) lead-in/lead-out tracks
- Limit to track length, not absolute LBA of next track

Previously, Saturn and Sega CD discs are all hashed to the same value `D41D8CD98F00B204E9800998ECF8427E` (empty sequence). This happens because the method is trying to hash the non-existent lead-in track and `track.NextTrack.LBA` is `0` (the LBA of the "real" first track).

Hashing PC Engine discs already mostly worked because the first track is usually an audio track and the lead-in track copies the mode of the first track and is skipped. A handful of discs where the first data track is <512 sectors may generate a different hash now, because the old implementation was reading past the end of the track. The only examples of that I could find were some unlicensed games that weren't in the gamedb.

Jaguar CD hashing still doesn't work because all tracks are marked as audio on those discs.

[//]: # "Apart from the mandatory license signature, these tasks are optional, but doing them could save reviewers some time and get the PR merged sooner."
Check if completed:
- [x] I have run any relevant test suites
- [x] I, the committer, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-06-22) and am compliant

### CODE REVIEW

**Comment by YoshiRulz** [comment]
Does this resolve #1075 then? That would be huge.

**Comment by kalimag** [comment]
Kind of? I tested a fair number of Saturn, Sega CD and PC Engine discs, and didn't find any collisions between different games. However, different revisions/bad dumps do occasionally produce the same hash, which is expected since it's only hashing the first track.

BizHawk now has another set of console-specific disc hashing functions in the RA code. I don't know how well they deal with revisions, but perhaps reusing those for general hashing would also be an option. They currently live in the `EmuHawk` project, but other than swapping out the RA `ConsoleID` for BizHawks `DiscId` or `SystemId` it probably wouldn't be hard to move them into the disc or common project.

**Comment by CasualPokePlayer** [comment]
> BizHawk now has another set of console-specific disc hashing functions in the RA code. I don't know how well they deal with revisions, but perhaps reusing those for general hashing would also be an option. They currently live in the EmuHawk project, but other than swapping out the RA ConsoleID for BizHawks DiscType or SystemId it probably wouldn't be hard to move them into the disc or common project.

FWIW, doing this could also help solve the issue of 3DS hashing, as rcheevos has their own limited hashing method for it. The downside of course, is it requires firmware (aes_keys.txt / seeddb.bin) in the case the game is encrypted (although if the firmware is not present the game couldn't be loaded anyways).

The disc hashing code too is also meant to be more a secondary implementation compared to the rcheevos implementation, although I've been mulling considering just removing it and deferring the hashing code entirely to rcheevos, like debug hashing code does (it has callbacks set in place so it will end up using our own disc system just fine). For general hashing however, it could perhaps still be better to have our own implementation so we are not directly subject to rcheevos in normal use (i.e. most users).

**Comment by CasualPokePlayer** [comment]
> Does this resolve #1075 then? That would be huge.

Jaguar CD still won't work with this approach. Fundementally, Jaguar CDs are "audio" CDs, they never have "data tracks" (this was a design decision by Atari to squeeze out more space, at of course the obvious cost of no error correction data and so a much higher risk of disc read failures). You'd want a specialized disc hashing approach here, likely doing what [RA does](https://github.com/TASEmulators/BizHawk/blob/a72b0f7e8c181f3b140ab401238c4f77f08ad0f7/src/BizHawk.Client.EmuHawk/RetroAchievements/RetroAchievements.GameVerification.cs#L168-L275).

File: src/BizHawk.Emulation.DiscSystem/DiscHasher.cs [hunk 1]
```diff
@@ -87,12 +87,15 @@ public string OldHash()
 		{
 			var buffer = new byte[512 * 2352];
 			var dsr = new DiscSectorReader(disc);
-			foreach (var track in disc.Session1.Tracks)
+			// don't hash generated lead-in and lead-out tracks
+			for (int i = 1; i < disc.Session1.Tracks.Count - 1; i++)
```
Comment (CasualPokePlayer) [inline src/BizHawk.Emulation.DiscSystem/DiscHasher.cs:91]: Use `i <= disc.Session1.InformationTrackCount;` rather here.

**Review by CasualPokePlayer — CHANGES_REQUESTED** [review]

File: src/BizHawk.Emulation.DiscSystem/DiscHasher.cs (hunk 1 above)
Comment (YoshiRulz) [reply src/BizHawk.Emulation.DiscSystem/DiscHasher.cs:91]:
```suggestion
			for (int i = 1; i <= disc.Session1.InformationTrackCount; i++)
```
?

**Review by CasualPokePlayer — APPROVED** [review]

**Comment by kalimag** [comment]
- Is there any reason not to also hash the TOC for non-PSX discs? (apart from #3760)
- Should there be special handling for old movies saved with the meaningless `D41D8CD98F00B204E9800998ECF8427E` hash? 
- Is it alright to just duplicate the Jaguar CD hashing code from the RA integration into the DiscSystem project to cover that system as well?

**Comment by YoshiRulz** [comment]
I'm afraid this didn't end up making it into RC1. I think it's because I pushed `master:release` without pulling `master`.

**Pull request merged**